# Short NumPy Tutorial

* NumPy is a Python package for numerical computation. 
* The core data type is the array.
* NumPy functions operate on arrays.

NumPy (np) supports creating arrays with different characteristics
* **np.array([list of numbers])** - array from a list
* **np.arange(start, stop, step)** - array whose entries are from start to stop and each entry is incremented by step (compared to the previous entry)
* **np.zeros((rows, cols))** - array of zeros of shape (rows, cols)
* **np.ones((rows, cols))** - array of zeros of shape (rows, cols)
* **np.random.randint(start, stop, num)** -- array of num entries randomly chosen from start to stop
* **np.random.rand(num)** array of random num float entries from 0 to 1

as well as a wide variety of arithmetic operations on arrays
* **np.add()** - Addition (e.g., 1 + 1 = 2)
* **np.subtract()** - Subtraction (e.g., 3 - 2 = 1)
* **np.negative()** - Unary negation (e.g., -2)
* **np.multiply()** -	Multiplication (e.g., 2 * 3 = 6)
* **np.divide()**  -	Division (e.g., 3 / 2 = 1.5)
* **np.floor_divide()** -	Floor division (e.g., 3 // 2 = 1)
* **np.power()** - 	Exponentiation (e.g., 2 ** 3 = 8)
* **np.mod()** - 	Modulus/remainder (e.g., 9 % 4 = 1)

and standard trigonometry functions
* **np.sin()** - sin
* **np.cos()** - cosine
* **np.tan()** - tangent
* **np.asin()** - arc sine
* **np.acos()** - arc cosine
* **np.atan()** - arc tangent
* **np.hypot()** - given the sides of a triangle, returns hypotenuse

Note: this is a small fraction of the functions provided by NumPy. For more information see:
* https://numpy.org/doc/stable/user/tutorials_index.html
* https://www.w3schools.com/python/numpy_intro.asp

#### Note: You can also write functions in Jupyter Notebooks

In [1]:
def hello(name):
  print ("Hello ", name)

hello("DAT1")

Hello  DAT1


#### Installing NumPy

Option A: **Using Anaconda**

In [2]:
#type in a terminal
#conda install numpy

Option B: **Using pip**

In [3]:
#type in a terminal
#pip install numpy

#### Using NumPy


In [4]:
#importing numpy
import numpy as np

In [5]:
#list with % of low emission Licensed Taxis operating in York per year from 2011 to 2019
yorkTaxis = [0.00, 1.00, 3.00, 6.00, 8.00, 13.00, 16.10, 17.60, 19.90]

#list with % of low emission Licensed Taxis operating in Leeds per year from 2011 to 2019
leedsTaxis = [0.50, 1.20, 2.30, 5.00, 7.80, 14.30, 17.10, 17.30, 20.90]


In [6]:
#convert lists into numpy array
york  = np.array(yorkTaxis)
leeds = np.array(leedsTaxis)

In [7]:
#print the shape of the arrays
print(york.shape)
print(leeds.shape)

(9,)
(9,)


In [8]:
#print the data types of the arrays
print(york.dtype)
print(leeds.dtype)

float64
float64


In [9]:
#get the values for York and Leeds from 2015 until 2019
#note: index 4 corresponds to 2015
print(york[4:])
print(leeds[4:])

[ 8.  13.  16.1 17.6 19.9]
[ 7.8 14.3 17.1 17.3 20.9]


In [10]:
#find the min, max, and mean of each city
yorkMin  = np.min(york)
yorkMax  = np.max(york)
yorkMean = np.mean(york)
print(yorkMin, yorkMax, yorkMean)

leedsMin  = np.min(leeds)
leedsMax  = np.max(leeds)
leedsMean = np.mean(leeds)
print(leedsMin, leedsMax, leedsMean)

0.0 19.9 9.399999999999999
0.5 20.9 9.600000000000001


In [11]:
#concatenate the two arrays
allData = np.concatenate((york, leeds))
print(allData)

[ 0.   1.   3.   6.   8.  13.  16.1 17.6 19.9  0.5  1.2  2.3  5.   7.8
 14.3 17.1 17.3 20.9]


In [12]:
#sort the concatenated array
sortedAllData = np.sort(allData)
print(sortedAllData)

[ 0.   0.5  1.   1.2  2.3  3.   5.   6.   7.8  8.  13.  14.3 16.1 17.1
 17.3 17.6 19.9 20.9]


In [13]:
#transform the concatenated array into a 2D array where each column holds values for each city
allData2D = allData.reshape(9, 2)
print(allData2D)

[[ 0.   1. ]
 [ 3.   6. ]
 [ 8.  13. ]
 [16.1 17.6]
 [19.9  0.5]
 [ 1.2  2.3]
 [ 5.   7.8]
 [14.3 17.1]
 [17.3 20.9]]


In [14]:
#transpose the allData2D array
print(allData2D.T)

[[ 0.   3.   8.  16.1 19.9  1.2  5.  14.3 17.3]
 [ 1.   6.  13.  17.6  0.5  2.3  7.8 17.1 20.9]]


## A few tasks on using NumPy 

**T1) Find the difference in % of low emission licensed taxis between York and Leeds each year**

In [15]:
#Write your code here
diff = np.subtract(york,leeds)
print(diff)

[-0.5 -0.2  0.7  1.   0.2 -1.3 -1.   0.3 -1. ]


In [16]:
#This also works because york and leeds are numpy arrays
print(york-leeds)

#The following does not work because both yorkTaxis and leedsTaxis are lists
#yorkTaxis-leedsTaxis 

[-0.5 -0.2  0.7  1.   0.2 -1.3 -1.   0.3 -1. ]


**T2) Complete the function minMaxDiff which returns the minimum and maximum differences in % of low emission licensed taxis between York and Leeds.**

In [17]:
#Write your code here
def minMaxDiff(diff):
  return np.min(diff), np.max(diff)

minMaxDiff(diff)

(-1.3000000000000007, 1.0)

**T3) Write a function that counts how many years York had more taxis with lower emission than Leeds**

Hint: you may want to look at the NumPy functions _count_nonzero_ or _where_

In [18]:
#Write your code here
def countEmissions(a):
  return np.count_nonzero(a > 0)

def countEmissions2(a):
  return np.size(np.where (a > 0))

print(countEmissions(diff))
print(countEmissions2(diff))

4
4


**T4) Using the data for York taxis, create a 2D array in which each entry from the york array is associated with its year, e.g., [[2011, 2012, ...][0.00, 1.00, ...]]**

Hints: 
* use the NumPy commands for array creation from above
* check the vstack Numpy function


In [19]:
#Write your code here
years = np.arange(2011, 2020,1)
yorkData = np.vstack((years, york))
yorkData

array([[2.011e+03, 2.012e+03, 2.013e+03, 2.014e+03, 2.015e+03, 2.016e+03,
        2.017e+03, 2.018e+03, 2.019e+03],
       [0.000e+00, 1.000e+00, 3.000e+00, 6.000e+00, 8.000e+00, 1.300e+01,
        1.610e+01, 1.760e+01, 1.990e+01]])

In [20]:
#This also works
years = np.arange(2011, 2020,1)
yorkData = np.array([years, york])
yorkData

array([[2.011e+03, 2.012e+03, 2.013e+03, 2.014e+03, 2.015e+03, 2.016e+03,
        2.017e+03, 2.018e+03, 2.019e+03],
       [0.000e+00, 1.000e+00, 3.000e+00, 6.000e+00, 8.000e+00, 1.300e+01,
        1.610e+01, 1.760e+01, 1.990e+01]])

**Notes** 

* NumPy supports (primarily) homogeneous arrays (i.e., arrays of the same datatype). This is why the years are in the following format: $2.011e+03 = 2.011 x 10^3 = 2011$. 
* NumPy arrays can support heteregenous datatypes (see the next practical) but this is not ideal. Hence why in the following practicals we will switch to the Pandas library that has native support for manipulating datasets with heterogenous datatypes.

#### Read Chapter 2 - Introduction to NumPy from the Python Data Science Handbook
https://jakevdp.github.io/PythonDataScienceHandbook/02.00-introduction-to-numpy.html